spi/max7219_32x8_spi/max7219_32x8_spi.c

```diff
@@ -0,0 +1,147 @@
+    gpio_put(PICO_DEFAULT_SPI_CSN_PIN, 0);  // Active low
+    asm volatile("nop \n nop \n nop");
+}
+
+static inline void cs_deselect() {
+    asm volatile("nop \n nop \n nop");
+    gpio_put(PICO_DEFAULT_SPI_CSN_PIN, 1);
+    asm volatile("nop \n nop \n nop");
+}
+#endif
+
+#if defined(spi_default) && defined(PICO_DEFAULT_SPI_CSN_PIN)
+static void write_register(uint8_t reg, uint8_t data) {
+    uint8_t buf[2];
+    buf[0] = reg;
+    buf[1] = data;
+    cs_select();
+    spi_write_blocking(spi_default, buf, 2);
+    cs_deselect();
+    sleep_ms(1);
+}
+
+static void write_register_all(uint8_t reg, uint8_t data) {
+    uint8_t buf[2];
+    buf[0] = reg;
+    buf[1] = data;
+    cs_select();
+    for (int i = 0; i< NUM_MODULES;i++)
+        spi_write_blocking(spi_default, buf, 2);
+    cs_deselect();
+}
+
+#endif
+
+int main() {
+    stdio_init_all();
+
+#if !defined(spi_default) || !defined(PICO_DEFAULT_SPI_SCK_PIN) || !defined(PICO_DEFAULT_SPI_TX_PIN) || !defined(PICO_DEFAULT_SPI_RX_PIN) || !defined(PICO_DEFAULT_SPI_CSN_PIN)
+#warning spi/bme280_spi example requires a board with SPI pins
+    puts("Default SPI pins were not defined");
+#else
+
+    printf("Hello, max7219! Drawing things on a LED matrix since 2022...\n");
+
+    // This example will use SPI0 at 10MHz.
+    spi_init(spi_default, 10 * 1000 * 1000);
+    gpio_set_function(PICO_DEFAULT_SPI_RX_PIN, GPIO_FUNC_SPI);
+    gpio_set_function(PICO_DEFAULT_SPI_SCK_PIN, GPIO_FUNC_SPI);
+    gpio_set_function(PICO_DEFAULT_SPI_TX_PIN, GPIO_FUNC_SPI);
+
+    // Make the SPI pins available to picotool
+    bi_decl(bi_3pins_with_func(PICO_DEFAULT_SPI_RX_PIN, PICO_DEFAULT_SPI_TX_PIN, PICO_DEFAULT_SPI_SCK_PIN, GPIO_FUNC_SPI));
+
+    // Chip select is active-low, so we'll initialise it to a driven-high state
+    gpio_init(PICO_DEFAULT_SPI_CSN_PIN);
+    gpio_set_dir(PICO_DEFAULT_SPI_CSN_PIN, GPIO_OUT);
+    gpio_put(PICO_DEFAULT_SPI_CSN_PIN, 1);
+
+    // Make the CS pin available to picotool
+    bi_decl(bi_1pin_with_name(PICO_DEFAULT_SPI_CSN_PIN, "SPI CS"));
+
+    // Send init sequence to device
+
+    write_register_all(CMD_SHUTDOWN, 0);
+    write_register_all(CMD_DISPLAYTEST, 0);
+    write_register_all(CMD_SCANLIMIT, 7);  // Use all lines
+    write_register_all(CMD_DECODEMODE, 0); // No BCD decode, just use bit pattern.
+    write_register_all(CMD_SHUTDOWN, 1);
+    write_register_all(CMD_BRIGHTNESS, 8);
+
+    for (int i=0; i<256; i++)
+    {
+        for (int j=0; j<8; j++)
+
+        write_register_all(CMD_DIGIT0+j, i);
+	    sleep_ms(20);
+    }
+
+    int bright = 1;
+    while (1)
+    {
+        for (int i=0; i<8; i++)
+        {
+            if (bright & 1)
+            {
+                write_register_all(CMD_DIGIT0 + i, 170 >> (i%2));
+            }
+            else
+            {
+                write_register_all(CMD_DIGIT0 + i, (170 >> 1) << (i%2));
+            }       
+
+	        write_register_all(CMD_BRIGHTNESS, bright % 16);
+        }
+	    sleep_ms(250);
```

Comment:
Messed up indentation due to mix of spaces and tabs? (IIRC we generally uses spaces-only in the other examples)